EasyPost/Hooks.cs

```diff
@@ -0,0 +1,135 @@
+using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+
+namespace EasyPost;
+
+public class Hooks
+{
+    /// <summary>
+    ///     An <see cref="EventHandler{OnRequestExecutingEventArgs}"/> to view an HTTP request by the client prior to being sent.
+    ///     Editing the <see cref="HttpRequestMessage"/> in this callback does not impact the <see cref="HttpRequestMessage"/> being executed.
+    /// </summary>
+    public EventHandler<OnRequestExecutingEventArgs>? OnRequestExecuting { get; set; }
+
+    /// <summary>
+    ///     An <see cref="EventHandler{OnRequestResponseReceivedEventArgs}"/> to view an HTTP response received by the client.
+    /// </summary>
+    public EventHandler<OnRequestResponseReceivedEventArgs>? OnRequestResponseReceived { get; set; }
+}
+
+public class OnRequestExecutingEventArgs : EventArgs
+{
+    /// <summary>
+    ///     The <see cref="HttpRequestMessage"/> about to be executed by the HTTP request.
+    /// </summary>
+    private HttpRequestMessage Request { get; } // Not publicly-exposed, used for reference only.
+
+    /// <summary>
+    ///     The <see cref="HttpMethod"/> of the HTTP request.
+    /// </summary>
+    public HttpMethod RequestMethod => Request.Method;
+
+    /// <summary>
+    ///     The <see cref="Uri"/> of the HTTP request.
+    /// </summary>
+    public Uri? RequestUri => Request.RequestUri;
```

Comment:
Question: why are there `?`

Reply:
Can be null

Reply:
The URI should never be null when making an API request, can we remove this?

Reply:
Not without the compiler complaining. We know it will never be null, but by the typing of the underlying `HttpRequestMessage` class, it COULD be null.
<img width="584" alt="image" src="https://github.com/EasyPost/easypost-csharp/assets/17054780/76e6b2ab-c696-47ea-8d89-ea87d60ec29b">